sklearn/multilabel/classifier_chain.py

```diff
@@ -0,0 +1,134 @@
+        """
+        self.base_estimator = base_estimator
+        self.random_state = random_state
+        if order is not None:
+            if not isinstance(order, list):
+                order = list(order)
+            if not all([isinstance(i, int) for i in order]):
+                raise ValueError("all elements of order parameter must be "
+                                 "integers")
+
+        self.order = order
+
+        self.shuffle = shuffle
+
+    def fit(self, X, Y):
+        """Fit the ClassifierChain model according to the given training (X)
+        and the true labels (Y) of classes that appear earlier in the chain as
+        defined by the order parameter.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        Y : array, shape (n_samples, n_classes)
+
+        Attributes
+        ----------
+        classifiers_ : list
+            A list of copies of base_estimator. Once fit the classifiers in
+            this list will be ordered as specified by the order attribute.
+            The original order of labels (as specified by Y) is recovered in
+            the predict method by indexing into the predictions with the
+            list of indices in the order attribute.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+
+        if issparse(Y):
+            Y = Y.toarray()
+
+        random_state = check_random_state(self.random_state)
+        self.classifiers_ = [clone(self.base_estimator)
+                             for _ in range(Y.shape[1])]
+
+        if self.order is not None:
+            if not len(self.order) == Y.shape[1]:
+                raise ValueError("chain_order length must equal n_labels")
+        else:
+            self.order = list(range(Y.shape[1]))
+
+        if self.shuffle:
+            random_state.shuffle(self.order)
+
+        for chain_idx, classifier in enumerate(self.classifiers_):
+            previous_labels = Y[:, self.order[:chain_idx]]
+            y = Y[:, self.order[chain_idx]]
+            X_aug = np.hstack((X, previous_labels))
+            classifier.fit(X_aug, y)
+
+    def predict(self, X):
+        """For each row in X make a prediction for each of the models in
+        the chain using the features available in X plus the predictions of
+        previous models in the chain as defined by the order parameter.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+
+        Returns
+        -------
+        Y_pred : array, shape (n_samples, n_classes)
+        """
+
+        Y_pred_chain = np.zeros((X.shape[0], len(self.classifiers_)))
+        for chain_idx, classifier in enumerate(self.classifiers_):
+            previous_predictions = Y_pred_chain[:, :chain_idx]
+            X_aug = np.hstack((X, previous_predictions))
+            Y_pred_chain[:, chain_idx] = classifier.predict(X_aug)
```

Comment:
For classification, we might want to be able to incorporate not just the categorical predictions of the previous model, but its `predict_proba` or `decision_function` downstream in the chain. We can leave this feature until version 2.0 (just as we might leave multioutput regression to version 2.0, but I'd like to see that)